### DISCUSSION TITLE
CUE and Canonical JSON

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#1285] [General]
Canonical JSON creates deterministic ordering on JSON documents, primarily used for consistent hashing, so that equivalent documents are "sorted" resulting in the same signature or hash.

- http://wiki.laptop.org/go/Canonical_JSON
- https://tools.ietf.org/id/draft-rundgren-json-canonicalization-scheme-05.html
- https://www.rfc-editor.org/rfc/rfc8785
- https://github.com/gibson042/canonicaljson-go
- https://hackage.haskell.org/package/canonical-json

This topic came up in https://github.com/in-toto/in-toto-golang/issues/129, can CUE validate that a JSON document is in canonical form. I wonder if it also applies to Kubernetes ConfigMaps where the contents of the data section are created by CUE. In theory Kubernetes could see a different ConfigMap despite having the "same" contents.

CUE does not have any guarantees on field order, though they tend to be topological in practice. It could be useful to have a canonical output mode for when signing or hashing is required.

**Comment by joulaud** [comment]
Something like that would be more than useful. Even in development process I want to be able to compare export or eval results at each change.  I use a pipe through `jq  --sort-keys` to obtain some sort of stability but it does not cover every case.

**Comment by eadlam** [comment]
This would be great. I like to commit materialized json files to git, and non-deterministic order ruins diffs. I work around it with graphql, but would rather not have to.

**Reply by verdverm** [reply]
It's more of a topological sort. It seems deterministic, that given the same inputs, the same outputs are generated. Agreed that small changes to CUE can have significant diffs in a json or yaml export.
